docs/src/components/DocNavigation/index.tsx

```diff
@@ -0,0 +1,190 @@
+import React from "react";
+import { useStaticQuery, graphql } from "gatsby";
+
+import Wrapper, { WrapperDesktop, WrapperMobile } from "./primitives/Wrapper";
+import Collapse from "./primitives/Collapse";
+import Category from "./primitives/Category";
+import Link from "./primitives/Link";
+import useDevMode from "../../hooks/useDevMode";
+
+type Trail = Array<{
+  name: string;
+  url: string;
+}>;
+
+interface QueryData {
+  allMdx: {
+    nodes: Array<{
+      id: string;
+      fields: {
+        tabCollection: string | null;
+        trail: Trail;
+      };
+      parent: {
+        name: string;
+      };
+    }>;
+  };
+}
+
+type NavigationItem =
+  | {
+      type: "branch";
+      name: string;
+      items: Navigation;
+    }
+  | {
+      type: "leaf";
+      name: string;
+      url: string;
+    };
+
+type Navigation = NavigationItem[];
+
+export function groupTrails(trails: Trail[]): Navigation {
+  // https://stackoverflow.com/a/57344801/1247274
+  const result: Navigation = [];
+  const level = { result };
+  trails.forEach(trail => {
+    trail.reduce((acc, { name, url }, index) => {
+      if (!acc[name]) {
+        acc[name] = { result: [] };
+        if (index < trail.length - 1) {
+          acc.result.push({ type: "branch", name, items: acc[name].result });
+        } else {
+          acc.result.push({ type: "leaf", name, url });
+        }
+      }
+      return acc[name];
+    }, level);
+  });
+  return result;
+}
+
+function getItemKey(item: NavigationItem) {
+  if (item.type === "branch") {
+    return `${item.type}-${item.name}`;
+  }
+  return item.url;
+}
+
+function DocNavigationItem({
+  devMode,
+  currentUrl,
+  level,
+  item,
+}: {
+  devMode: boolean;
+  currentUrl: string;
+  level: number;
+  item: NavigationItem;
```

Comment:
can you create an interface and pass it here? should look better :) 
you can create a separate file inside the folder `DocNavigation` called `records` or `interfaces` and pass all types there, so this file will look more cleaner, just a tip.